utils/ada_helpers.py

```diff
@@ -0,0 +1,451 @@
+            return self.__dict[k]
+        return 0
+
+    def assets(self) -> List[str]:
+        keys = []
+        for k, v in self.__dict.items():
+            keys.append(k)
+        return keys
+
+    def size(self) -> int:
+        return len(self.__dict)
+
+    def is_empty(self) -> bool:
+        return self.size() == 0
+
+
+class CardanoGQLClient:
+    def __init__(self, gql_server_url: str, server_caller: Union[Callable, None]):
+        self.__gql_server_url = gql_server_url
+        if server_caller is None:
+            self.__gql_server_caller = CardanoGQLClient.__default_client_caller
+        else:
+            self.__gql_server_caller = server_caller
+
+
+    def get_current_slot(self) -> int:
+        query = r'''query { cardano { tip { slotNo } } }'''
+        result = self.__gql_server_caller(self.__gql_server_url, query)
+        if 'cardano' in result and 'tip' in result['cardano'] and 'slotNo' in result['cardano']['tip']:
+            return result['cardano']['tip']['slotNo']
+        else:
+            raise Exception(f'Bad response from GQL: {result}')
+
+    def submit_tx(self, signed_tx: bytes) -> str:
+        query = r'''mutation SendRawTx($rawtx: String!) {submitTransaction(transaction: $rawtx) {hash}}'''
+        result = self.__gql_server_caller(self.__gql_server_url, query, {'rawtx': signed_tx.hex()})
+        if 'submitTransaction' in result and 'hash' in result['submitTransaction']:
+            return result['submitTransaction']['hash']
+        else:
+            raise Exception(f'Bad response from gql: {result}')
+
+    def get_utxos(self, addresses: List[CardanoAddress], native_only: bool) -> List[CardanoUTxO]:
+        query = r'''query utxoSetForAddresses ( $addresses: [String]!) { utxos( where: { address: { _in: $addresses }} ) { address transaction { block { number } } tokens { asset { name ticker } quantity } txHash index value } }'''
+
+        addresses_str_arr = []
+        addresses_index_map = dict()
+        for address in addresses:
+            addresses_str_arr.append(address.address)
+            addresses_index_map[address.address] = address.derivation_index
+
+        result = self.__gql_server_caller(self.__gql_server_url, query, {'addresses': addresses_str_arr})
+
+        if 'utxos' not in result:
+            raise Exception(f'Bad response from GQL: {result}')
+
+        utxos = []
+        for utxo in result['utxos']:
+            if 'transaction' in utxo and 'block' in utxo['transaction'] and 'number' in utxo['transaction']['block'] \
+                    and 'txHash' in utxo and 'index' in utxo and 'address' in utxo and 'value' in utxo:
+                has_tokens = 'tokens' in utxo and (len(utxo['tokens']) > 0)
+                if has_tokens and native_only:
+                    continue
+
+                utxo_address = utxo['address']
+                utxo_tokens = []
+                for token in utxo['tokens']:
+                    if 'quantity' in token and 'asset' in token \
```

Comment:
Please validate if has_tokens = false before,
if utxo['tokens'] doesn't exists we will throw

Reply:
:+1: